[SYCL][Matrix spec] Add joint_matrix_prefetch and overloads of load/store with annotated_ptr

dkhaldi opened the pull request.

dkhaldi commented:
@gmlueck, gentle ping to review the spec of overloads of load/store on annotated pointer and prefetch.

gmlueck reviewed (commented):
It seems like we should have some documentation about  which properties can be used with `joint_matrix_load` and `joint_matrix_store`.  The same goes for `joint_matrix_prefetch`.  That function takes a property list, but there's no documentation about which properties you can pass.

dkhaldi commented:
> It seems like we should have some documentation about which properties can be used with `joint_matrix_load` and `joint_matrix_store`. The same goes for `joint_matrix_prefetch`. That function takes a property list, but there's no documentation about which properties you can pass.

I added some documentation as we will use existing properties for cache control and cache levels. Is what I added enough?

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
````diff
@@ -322,11 +322,126 @@ q.submit([&](sycl::handler& cgh) {
 q.wait();
 ```
+
+=== Load/Store Using Annotated Pointers
````
I think we could define these load/store variants with `annotated_ptr` in the "oneapi" spec because there could be some "oneapi" properties that make sense.  For example, there is another conversation where we think `alignment` would be useful.  If we do that, this "intel" spec can document the "intel" properties that can also be used with these APIs.

dkhaldi replied:
@gmlueck, that makes sense. 
But are you suggesting adding these at the end of the "oneapi" document or blend them with the existing ones under the same section?  
I am hesitant about adding/mixing these new APIs with the ones we already have, that are somewhat stable. In the event we decide to promote the stable parts, there should be clear distinction between the stable ones and the new added ones that should also go through the cycle of implementation and stable usage.
Do you think there is a good way to address this?

gmlueck replied:
I was thinking that they would be documented in the same section as the other load/store APIs.

When it comes time to promote the matrix extension out of "experimental", we can decide whether these APIs are not yet ready.  If that is that case, we can remove them from the "supported" extension and leave them in the "experimental" one.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -322,11 +322,126 @@ q.submit([&](sycl::handler& cgh) {
+using syclex = sycl::ext::oneapi::experimental;
+using syclintelex = sycl::ext::intel::experimental;
+
+auto aptr = annotated_ptr(ptr, syclex::properties{syclintelex::cache_control_read<cache_control_read_type::cached, cache_level::L2>});
```
Keep an eye on #11584, which is changing the syntax for these cache controls.  You will need to update this example when that PR stabilizes.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
````diff
@@ -322,11 +322,126 @@ q.submit([&](sycl::handler& cgh) {
+});
+```
+
+=== Prefetch
````
The "sycl_ext_oneapi_prefetch" extension is a "oneapi" extension.  Therefore, can't we put this `joint_matrix_prefetch` in the "oneapi" part of the joint matrix extension?

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -322,11 +322,126 @@ q.submit([&](sycl::handler& cgh) {
+The level of cache targeted by `joint_matrix_prefetch` in the last
+argument is specified using the compile-time properties defined in the
+SYCL extension
+link:../../proposed/sycl_ext_oneapi_prefetch.asciidoc[sycl_ext_oneapi_prefetch.asciidoc#specifying-cache-levels]
```
This link doesn't work in the HTML render.  I think the "#specifying-cache-levels" part needs to be outside of the square brackets.

dkhaldi commented:
@gmlueck, as we discussed on Monday, I put the prefetch size as template arguments since they must be constexpr for the implementation.  
Can you please review/approve?

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -248,6 +267,33 @@ fashion. `stride` describes the number of elements between consecutive
+using syclintelex = sycl::ext::intel::experimental;
+
+auto A_ptr = syclex::annotated_ptr{A,
+               syclex::properties{syclintelex::read_assertion<
+                   syclintelex::cache_control<syclintelex::cache_mode::constant,
+                                              syclex::cache_level::L2>>}};
```
```suggestion
               syclex::properties{syclintelex::read_hint<
                   syclintelex::cache_control<syclintelex::cache_mode::cached,
                                              syclex::cache_level::L2>>}};
```

The comment below says "this load will be cached to L2".  Assuming that's what you want the example to show, then you want `read_hint` here, not `read_assertion`.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -258,6 +304,13 @@ void joint_matrix_store(Group g,
    const joint_matrix<Group, T, use::accumulator, Rows, Cols, layout::dynamic> &res,
    multi_ptr<T, Space, IsDecorated> dest, size_t stride, layout Layout);
 
+template <typename Group, typename T1, typename T2, size_t Rows, size_t Cols,
+          typename PropertyListT>
+void joint_matrix_store(Group g,
+   const joint_matrix<Group, T1, use::accumulator, Rows, Cols, layout::dynamic> &res,
+   ext::oneapi::experimental::annotated_ptr<T2, PropertyListT> dest,
+   size_t stride, layout Layout);
```
```suggestion
template <typename Group, typename T, size_t Rows, size_t Cols,
          typename PropertyListT>
void joint_matrix_store(Group g,
   const joint_matrix<Group, T, use::accumulator, Rows, Cols, layout::dynamic> &res,
   ext::oneapi::experimental::annotated_ptr<T, PropertyListT> dest,
   size_t stride, layout Layout);
```

The other overload of `joint_matrix_store` has only one `T` template parameter (not two), so I presume you want the same here.

Note, however, in another PR I suggested changing `joint_matrix_store` to have two template parameters `T1` and `T2`.  So, I do think it makes sense for *both* overloads to have these two template parameters.  However, I think you  should make that change in a separate PR, which changes both overloads.

dkhaldi replied:
You are right. Let's add T1 and T2 in the other PR https://github.com/intel/llvm/pull/12276
Whichever PR gets merged first, the other one will follow.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -371,6 +429,46 @@ joint_matrix_apply(sg, C, [=](T &x) {
+
+```c++
+using syclex = sycl::ext::oneapi::experimental;
+using syclintelex = sycl::ext::intel::experimental;
````
Delete this line.  `syclintelex` isn't used in this example.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -371,6 +429,46 @@ joint_matrix_apply(sg, C, [=](T &x) {
+argument is specified using the compile-time properties defined in the
+SYCL extension
+link:../../proposed/sycl_ext_oneapi_prefetch.asciidoc[sycl_ext_oneapi_prefetch.asciidoc#specifying-cache-levels]
+as illustrated in the example below:
```
I think you should also describe the behavior when there are no properties.  For example:

> ... as illustrated in the example below.  When no cache levels are specified, the default behavior is to prefetch into the lowest level cache (i.e. L1).
>
> [Example]

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -439,6 +537,22 @@ void joint_matrix_load(Group g,
+    joint_matrix<Group, precision::tf32, Use, Rows, Cols, Layout> &res,
+    ext::oneapi::experimental::annotated_ptr<float, PropertyListT> src,
+    size_t stride);
+
```
In the other PR I reviewed, I proposed a different way to handle these special TF32 overloads.  Notice that with that proposal, you wouldn't need to add these new TF32 overloads for `annotated_ptr` because you would just use the regular non-TF32 versions.

dkhaldi replied:
Yes, let's solve this in  https://github.com/intel/llvm/pull/12276.
I will keep it this way for now because this is how it is done for the other load. Whichever PR gets merged first, the other one will follow.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -157,6 +157,15 @@ void joint_matrix_store(Group g,
     multi_ptr<T, Space, IsDecorated> src, size_t stride);
 
+// Only available when Layout != layout::dynamic
+template <typename Group, typename T1, typename T2,
```
Same comment here about `T1` and `T2`.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -228,6 +228,25 @@ void joint_matrix_load(Group g,
+          typename PropertyListT>
+void joint_matrix_load(Group g,
+    joint_matrix<Group, T1, use::accumulator, Rows, Cols, layout::dynamic> &res,
+    ext::oneapi::experimental::annotated_ptr<T2, PropertyListT> src,
```
Well, this function is defined in ext::oneapi::experimental::matrix namespace
annotated_ptr is defined in ext::oneapi::experimental namespace
Do you still think I should remove it?

gmlueck replied:
Yes.  This function declaration has visibility to identifiers defined in any of the parent namespaces, which includes `ext::oneapi::experimental`.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -157,6 +157,14 @@ void joint_matrix_store(Group g,
+void joint_matrix_store(Group g,
+    joint_matrix<Group, T, Use, Rows, Cols, Layout> &res,
+    ext::oneapi::experimental::annotated_ptr<T, PropertyListT> src,
+    size_t stride);
```
Yes for both questions. I am fixing this.

dkhaldi commented:
@intel/llvm-gatekeepers, can you please merge?
